daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala

```diff
@@ -0,0 +1,272 @@
+import com.daml.lf.value.Value
+import com.daml.lf.data.ImmArray
+
+import scala.annotation.tailrec
+
+private[lf] object NormalizeRollbacks {
+
+  type Nid = NodeId
+  type Cid = Value.ContractId
+  type TX = GenTransaction[Nid, Cid]
+  type Node = GenNode[Nid, Cid]
+  type LeafNode = LeafOnlyActionNode[Cid]
+  type ExeNode = NodeExercises[Nid, Cid]
+
+  // Normalize a transaction so rollback nodes satisfy the normalization rules.
+  // see `makeRoll` below
+
+  import Canonical.{Norm, Case, caseNorms}
+  // The normalization phase works by constructing intermediate values which are
+  // `Canonical` in the sense that only properly normalized nodes can be represented.
+  // Although this doesn't ensure correctness, one class of bugs is avoided.
+
+  import Trampoline.{Bounce, Land, Tramp}
+
+  def normalizeTx(txOriginal: TX): TX = {
+
+    // Here we traverse the original transaction structure.
+    // During the transformation, an original `Node` is mapped into a List[Norm]
+
+    // The List is necessary because the rules can:
+    // (1) drop nodes; (2) combine nodes (3) lift nodes from a lower level to a higher level.
+
+    txOriginal match {
+      case GenTransaction(nodesOriginal, rootsOriginal) =>
+        def traverseNids[R](xs: List[Nid])(k: List[Norm] => Tramp[R]): Tramp[R] = {
+          Bounce { () =>
+            xs match {
+              case Nil => k(Nil)
+              case x :: xs =>
+                traverseNode(nodesOriginal(x)) { norms1 =>
+                  traverseNids(xs) { norms2 =>
+                    Bounce { () =>
+                      k(norms1 ++ norms2)
+                    }
+                  }
+                }
+            }
+          }
+        }
+
+        def traverseNode[R](node: Node)(k: List[Norm] => Tramp[R]): Tramp[R] = {
+          Bounce { () =>
+            node match {
+
+              case NodeRollback(children) =>
+                traverseNids(children.toList) { norms =>
+                  makeRoll(norms)(k)
+                }
+
+              case exe: NodeExercises[_, _] =>
+                traverseNids(exe.children.toList) { norms =>
+                  k(List(Norm.Exe(exe, norms)))
+                }
+
+              case leaf: LeafOnlyActionNode[_] =>
+                k(List(Norm.Leaf(leaf)))
+            }
+          }
+        }
+        //pass 1
+        traverseNids(rootsOriginal.toList) { norms =>
+          //pass 2
+          pushNorms(initialState, norms) { (finalState, roots) =>
+            Land(GenTransaction(finalState.nodeMap, ImmArray(roots)))
+          }
+        }.run
+    }
+  }
+
+  // makeRoll: encodes the normalization transformation rules:
+  //   rule #1: ROLL [ ] -> ε
+  //   rule #2: ROLL [ ROLL [ xs… ] , ys… ] -> ROLL [ xs… ] , ROLL [ ys… ]
+  //   rule #3: ROLL [ xs… , ROLL [ ys… ] ] ->  ROLL [ xs… , ys… ]
+
+  //   rule #2/#3 overlap: ROLL [ ROLL [ xs… ] ] -> ROLL [ xs… ]
+
+  private def makeRoll[R](norms: List[Norm])(k: List[Norm] => Tramp[R]): Tramp[R] = {
```

Comment:
```suggestion
  private[this] def makeRoll[R](norms: List[Norm])(k: List[Norm] => Tramp[R]): Tramp[R] = {
```
almost only cosmetic

Reply:
done